### ISSUE TITLE
Compatible license with Elixir's one

### ISSUE THREAD

**Issue by eksperimental** [mykewould/crutches#84]
Hi, i just found out through the elixir mailing list about this project.
I have thought about something like this for a while, so kudos for creating something like this,
I have code to contribute, but before moving forward I would like to make a comment that can be taking into consideration.
Given the nature of the project, it is very likely that at some point some of the functions may end up making it into Elixir code base.
So i think to avoid conflicts in the future, it should be considered to release the code under a dual license, MIT/APACHE2, or maybe a public license.

I think if still be needed to release the rights to the elixir project if they want to be included as a special mention in the NOTICE file is unlikely to happen.

Maybe there's an easier solution, like anybody submitting code, agrees that it could potentially be added to Elixir project, maybe  we could include the credits in the commit message when added to elixir.

**Comment by druzn3k** [comment]
For me there's no problem. I would love to see pieces of code ported into the elixir codebase. :+1:

**Comment by eksperimental** [comment]
maybe it could be asked in the elixir mailing list what would be the best way to deal with a situation like this.
It can also help other projects in the same fashion as this one, where code depends on many contributors, not a single person.

**Comment by duijf** [comment]
I think this is a good discussion to start.

I've recently seen a few threads about relicensing code. It might be a bit of a painful job to do since we don't have a contributor licensing agreement, therefore we need permission of each individual contributor to relicense the project.

Let me see if I can find the relevant thread when I get home.

**Comment by duijf** [comment]
It was this issue: https://github.com/FLIF-hub/FLIF/issues/14

Main takeaway:

> If you wish to change the project's licensing terms at a later point, you need to have contributors release their contributions under the terms of a Contributor License Agreement starting now. Otherwise, at said point of future maturation, you'll have to get everybody who's ever contributed a patch to the project to explicitly sign off on your desired relicensing, which is a pain (as the maintainers of projects which have had to do this in the past, such as Bootstrap, can attest).

The whole thread is a pretty good discussion and overview of the issue. I don't have a clear idea how to move forward (not really versed in the legal matters, so I'll just leave the link here while I go read up).

**Comment by duijf** [comment]
Ok, so I did some reading and my understanding is the following (IANAL):
- When two software projects use different licenses, their licenses can be compatible or not.
- When one projects license is one directionally compatible with another, code can be shared that way.
- The MIT license (that we use) is one directionally compatible with the Apache 2.0 license (that Elixir uses).
- See the graphical representation below, if A uses the MIT license, B, which uses Apache 2.0 can use code from A without hassle (a comment explaining the source and attributing the creators would be enough).
- When you want to copy code from B to A, you need to jump through some hoops (luckily this isn't the case).

```
              compatible
A (uses MIT)  ----------> B (uses Apache 2.0)
```

So I guess for our code to be used in the Elixir project, nothing needs to be done. Functions can just be ported and relicensed under Apache 2.0. We _could_ ~~provide a dual-license~~ EDIT, meant: use a CLA, but that wouldn't make any sense, unless we want to go public domain at some point (which aparently has its own sort of issues).

See https://en.wikipedia.org/wiki/License_compatibility for more details.

**Takeaway:** I think no action is required for code from this repository to be included in Elixir's standard library.

**Comment by druzn3k** [comment]
Dude, you are a hero! :bow:

**Comment by duijf** [comment]
Haha thanks. Leaving this open for a while to collect comments.

**Comment by eksperimental** [comment]
@duijf thanks again for the write-up and research.. they are really useful!

Well my main point is that usually when you contribute code to elixir, you give your rights to the project.
that what happens with most code in github, because it will be impractical to maintain a list of who did what , specially when contributions are not whole pieces of code but corrections.

The thing is that I have the impression that if code from this project wants to make it into elixir, the project will have to resign the rights to the Elixir projects, unless they want to update their NOTICE file where the only one external source credited so far is Ericsson for one chuck of code that is used in the parser. That's why I recommended if this could be posted in the elixir mailing list to see what would be the best way. Maybe it wouldn't be an issue to add to the NOTICE file, what functions were originally taken from this project.
or the other proposition was to create a dual license (Apache 2.0 and MIT), and/or create an agreement that the code contributed to this project (and by the very nature of this project) may potentially make it to Elixir, and that the contributors.
I think it will be best to post it to the elixir-lang-talk list. to erase all doubts. I won't be active in the following days, so If somebody else can do it, it will be appreciated, otherwise I can do it when I'm free.
It will also help to other projects in a similar situation

**Comment by duijf** [comment]
@eksperimental Sorry for the delay. Thanks for sharing your viewpoint on this and providing the suggestions. I can start a new thread on the elixir-talk email list within a few days (if you want to do this, and can do it sooner, then feel free to do so).

If the Elixir project isn't willing to do this, then I don't think that adopting a dual license fixes the issue though. The MIT and Apache licenses are compatible according to the Wikipedia page I linked to. Basically the only condition of the MIT license is that attribution is granted (and it explicitly grants the right to relicense the code, as long as the attribution is there).

Currently, we do not have a CLA, so contributions are copyright their respective authors. By contributing, I'd argue they are agreeing with the conditions of the MIT license, which means that the Elixir project is free to adopt the code as long as the copyright notice is maintained. The Elixir project is also free to sub/relicense the code (as long as the attribution and the fact that the code is also available under the MIT license is kept).

The only thing that the Elixir project can't do is claim copyright of the code (which it would do if the NOTICE file isn't updated, and code from this project is merged in).

Basically we can go three ways:
1. Aks authors of functions that we'd like to see in the Elixir project to submit them on their own.
2. Submit the code to the Elixir project _on behalf of_ our contributors and ask that the NOTICE is updated crediting this project for a couple of functions (or add the attribution in a comment above the function).
3. Ask contributors to agree to having their code merged into the Elixir project and relinquish their copyright to Plataformatec when that happens: we can submit code to Elixir _on behalf of_ the contributors to Crutches. (This means adopting a CLA of some sort.)

Adopting a CLA makes us take on an administrative burden (each new contributor needs to 'sign' it or let us know that they agree with it). Option two makes the Elixir project take on that bit of administration. The first option makes both projects incur as little bureaucracy as possible&mdash;only when needed, not in the common case&mdash;and I think authors will be inclined to contribute their functions to the Elixir code base.

I think I prefer the first one. If an author does not respond to requests then we can always go with option two. If there are real issues with both approaches that I'm not seeing, then the third approach will be neccessary.

**Comment by eksperimental** [comment]
Erlang requires people to sign an agreement (i have to send an email myself, "signing" this agreement"),
but Elixir has a notice in the CONTRIBUTING file or somewhere saying that if you submit a PR you cede the rights to Plataforma tech. I think most of the projects is github are taking this latter approach.
So I guess Crutches could also do it.

Yes, please feel free to send the email, since you are an active maintainer of this project, I think it will make more sense.

regarding your three points,
Basically we can go three ways:

> 1. Aks authors of functions that we'd like to see in the Elixir project to submit them on their own.
>    Not feasible because many people can contribute to one set of functions in particular.
> 2. Submit the code to the Elixir project on behalf of our contributors and ask that the NOTICE is updated crediting this project for a couple of functions (or add the attribution in a comment above the function).
>    That's what we need to ask in the list.
> 3. Ask contributors to agree to having their code merged into the Elixir project and relinquish their copyright to Plataformatec when that happens: we can submit code to Elixir on behalf of the contributors to Crutches. (This means adopting a CLA of some sort.)
>    That's another option.

I think if Elixir team is willing to accept including third party functions and give credit to them, I think sticking to MIT would suffice.

another thing is that crediting the individual authors will be almost impossible to  track as the project grows specially for individual functions...
I think a more sensible approach since from what I can see this is a collective project, is to change the copyright ownership from a single person as it is now, to a collective one such as "The Crutches Project", but it beats me if weather that's legal or not.

thanks for looking into this @duijf

**Comment by duijf** [comment]
> Elixir has a notice in the CONTRIBUTING file or somewhere saying that if you submit a PR you cede the rights to Plataforma tech. I think most of the projects is github are taking this latter approach.
> So I guess Crutches could also do it.

I wouldn't know whether that would hold up. Sending an email would though, AFAIK. Going to look into that. If the approach Elixir uses holds up, then it seems like the best possible solution without us incurring a lot of bureaucracy.

> another thing is that crediting the individual authors will be almost impossible to track as the project grows specially for individual functions...

That is true and probably not ideal on the Elixir side. CLA seems best anyway.

>  I think a more sensible approach since from what I can see this is a collective project, is to change the copyright ownership from a single person as it is now, to a collective one such as "The Crutches Project", but it beats me if weather that's legal or not.

Depends on whether "The Crutches Project" is a legal entity (not until it we become some foundation or something).

I am currently thinking something along the lines of the following for the CLA:
1. Contributors agree to a CLA giving their copyright to some entity or person (in some manner).
2. The person getting the copyright gives the individual contributor a worldwide, perpetual, royalty free etc etc license to use the work (that was previously their own).
3. Because of 1 and 2, the person/entity that received the copyright is able to contribute the code to Elixir itself (attribution will not be needed).

Number 2 still needs some work/research. For instance, will the person who initially wrote the work and granted the copyright be bound by the terms of the MIT license or are they free to use that code however they like? Are they still able to submit it to the Elixir project themselves without action being taken by the person/entity that they granted the copyright to? Or would they be bound by the MIT license there too (so contributing their function to Elixir would mean having to attribute the person/entity that the copyright was granted to)?

Anyway, lot's to research and work out before I send that email. I hope I can find some time for that this week.

**Comment by eksperimental** [comment]
5+ year have passed
